# 2 Load digits / wind / breast_cancel 분류하기

## 2-1 Load digiets 손글씨 분류기

In [59]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier
import pandas as pd

# (2) 데이터 준비, data/label지정하기
digits = load_digits()
digits_data = digits.data
digits_label = digits.target

# digit데이터의 feature 출력해보기
print(digits.feature_names)
# digit데이터의 타겟 이름 출력해보기
print(digits.target_names)
# digit데이터의 키 이름 출력해보기
print(digits.keys())
# digit데이터의 describe 해보기
print(digits.DESCR+"\n")

# pandas로 출력해보기
digits_df = pd.DataFrame(data=digits_data, columns=digits.feature_names)
print("pandas------------------------------------")
print(digits_df)
print("구분------------------------------------")
      
      
# (3) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(digits_data, 
                                                    digits_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

# (4) 모델 학습 및 예측
# Desicion Tree
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)
print("Desicion Tree------------------------------------")
print(classification_report(y_test, y_pred))

#Random Forest 
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
print("Random Forest------------------------------------")
print(classification_report(y_test, y_pred))

#SVM Model  support Vector Machine
from sklearn import svm
svm_model = svm.SVC()
print(svm_model._estimator_type)

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)
print("SVM Model------------------------------------")
print(classification_report(y_test, y_pred))

#SGC Model Stochastic Gradient Descent Classifier
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
print(sgd_model._estimator_type)

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)
print("SGC Model------------------------------------")
print(classification_report(y_test, y_pred))


#Logistic Regression
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

print(logistic_model._estimator_type)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)
print("Logistic Regression------------------------------------")
print(classification_report(y_test, y_pred))

['pixel_0_0', 'pixel_0_1', 'pixel_0_2', 'pixel_0_3', 'pixel_0_4', 'pixel_0_5', 'pixel_0_6', 'pixel_0_7', 'pixel_1_0', 'pixel_1_1', 'pixel_1_2', 'pixel_1_3', 'pixel_1_4', 'pixel_1_5', 'pixel_1_6', 'pixel_1_7', 'pixel_2_0', 'pixel_2_1', 'pixel_2_2', 'pixel_2_3', 'pixel_2_4', 'pixel_2_5', 'pixel_2_6', 'pixel_2_7', 'pixel_3_0', 'pixel_3_1', 'pixel_3_2', 'pixel_3_3', 'pixel_3_4', 'pixel_3_5', 'pixel_3_6', 'pixel_3_7', 'pixel_4_0', 'pixel_4_1', 'pixel_4_2', 'pixel_4_3', 'pixel_4_4', 'pixel_4_5', 'pixel_4_6', 'pixel_4_7', 'pixel_5_0', 'pixel_5_1', 'pixel_5_2', 'pixel_5_3', 'pixel_5_4', 'pixel_5_5', 'pixel_5_6', 'pixel_5_7', 'pixel_6_0', 'pixel_6_1', 'pixel_6_2', 'pixel_6_3', 'pixel_6_4', 'pixel_6_5', 'pixel_6_6', 'pixel_6_7', 'pixel_7_0', 'pixel_7_1', 'pixel_7_2', 'pixel_7_3', 'pixel_7_4', 'pixel_7_5', 'pixel_7_6', 'pixel_7_7']
[0 1 2 3 4 5 6 7 8 9]
dict_keys(['data', 'target', 'frame', 'feature_names', 'target_names', 'images', 'DESCR'])
.. _digits_dataset:

Optical recognition of handwritte

Logistic Regression------------------------------------
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        43
           1       0.95      0.95      0.95        42
           2       0.98      1.00      0.99        40
           3       0.94      0.97      0.96        34
           4       0.97      1.00      0.99        37
           5       0.82      0.96      0.89        28
           6       1.00      0.96      0.98        28
           7       0.97      0.97      0.97        33
           8       0.92      0.81      0.86        43
           9       0.97      0.91      0.94        32

    accuracy                           0.95       360
   macro avg       0.95      0.95      0.95       360
weighted avg       0.95      0.95      0.95       360



/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


## 2-1 Digits 결론
Descision Tree : 0.86
Random Forest : 0.96
SVM :0.99
SGC : 0.94
Logistoc Regression : 0.95

SVM이 가장 높고 Desicision Tree가 가장 낮게 나왔다.
Desicision Tree는 가장 단순한 모델로 성능이 낮게 나오리라 예상은 했지만 나머지 모델들에 대해서는 사실 예상하기가 쉽지가 않다. 

## 2-2 wine 

In [60]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier

# (2) 데이터 준비, data/label지정하기
wine = load_wine()
wine_data = wine.data
wine_label = wine.target

# wine feature 출력해보기
print(wine.feature_names)
# wine 타겟 이름 출력해보기
print(wine.target_names)
# wine 키 이름 출력해보기
print(wine.keys())
# wine describe 해보기
print(wine.DESCR+"\n")

# pandas로 출력해보기
digits_df = pd.DataFrame(data=wine_data, columns=wine.feature_names)
print("pandas------------------------------------")
print(digits_df)
print("구분------------------------------------")
      

# (3) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(wine_data, 
                                                    wine_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

# (4) 모델 학습 및 예측
# Desicion Tree
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)
print("Desicion Tree------------------------------------")
print(classification_report(y_test, y_pred))

#Random Forest 
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
print("Random Forest------------------------------------")
print(classification_report(y_test, y_pred))

#SVM Model  support Vector Machine
from sklearn import svm
svm_model = svm.SVC()
print(svm_model._estimator_type)

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)
print("SVM Model------------------------------------")
print(classification_report(y_test, y_pred))

#SGC Model Stochastic Gradient Descent Classifier
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
print(sgd_model._estimator_type)

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)
print("SGC Model------------------------------------")
print(classification_report(y_test, y_pred))

#Logistic Regression
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

print(logistic_model._estimator_type)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)
print("Logistic Regression------------------------------------")
print(classification_report(y_test, y_pred))

['alcohol', 'malic_acid', 'ash', 'alcalinity_of_ash', 'magnesium', 'total_phenols', 'flavanoids', 'nonflavanoid_phenols', 'proanthocyanins', 'color_intensity', 'hue', 'od280/od315_of_diluted_wines', 'proline']
['class_0' 'class_1' 'class_2']
dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])
.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
       

/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


## 2-2 wine 결론
Descision Tree : 0.94 Random Forest : 1.00 SVM :0.61 SGC : 0.64 Logistoc Regression : 0.97

## 2-3 Breast Cancel

In [61]:
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import classification_report
from sklearn.ensemble import RandomForestClassifier

# (2) 데이터 준비, data/label지정하기
breast_cancer = load_breast_cancer()
breast_cancer_data = breast_cancer.data
breast_cancer_label = breast_cancer.target

# breast_cancer feature 출력해보기
print(breast_cancer.feature_names)
# breast_cancer 타겟 이름 출력해보기
print(breast_cancer.target_names)
# breast_cancer 키 이름 출력해보기
print(breast_cancer.keys())
# breast_cancer describe 해보기
print(breast_cancer.DESCR)

# pandas로 출력해보기
digits_df = pd.DataFrame(data=breast_cancer_data, columns=breast_cancer.feature_names)
print("pandas------------------------------------")
print(digits_df)
print("구분------------------------------------")
      

# (3) train, test 데이터 분리
X_train, X_test, y_train, y_test = train_test_split(breast_cancer_data, 
                                                    breast_cancer_label, 
                                                    test_size=0.2, 
                                                    random_state=7)

# (4) 모델 학습 및 예측
# Desicion Tree
decision_tree = DecisionTreeClassifier(random_state=32)
decision_tree.fit(X_train, y_train)
y_pred = decision_tree.predict(X_test)
print("Desicion Tree------------------------------------")
print(classification_report(y_test, y_pred))

#Random Forest 
from sklearn.ensemble import RandomForestClassifier

random_forest = RandomForestClassifier(random_state=32)
random_forest.fit(X_train, y_train)
y_pred = random_forest.predict(X_test)
print("Random Forest------------------------------------")
print(classification_report(y_test, y_pred))

#SVM Model  support Vector Machine
from sklearn import svm
svm_model = svm.SVC()
print("SVM Model------------------------------------")
print(svm_model._estimator_type)

svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)
svm_model.fit(X_train, y_train)
y_pred = svm_model.predict(X_test)
print("SVM Model------------------------------------")
print(classification_report(y_test, y_pred))

#SGC Model Stochastic Gradient Descent Classifier
from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
print(sgd_model._estimator_type)

sgd_model.fit(X_train, y_train)
y_pred = sgd_model.predict(X_test)
print("SGC Model------------------------------------")
print(classification_report(y_test, y_pred))

#Logistic Regression
from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()

print(logistic_model._estimator_type)

logistic_model.fit(X_train, y_train)
y_pred = logistic_model.predict(X_test)
print("Logistic Regression------------------------------------")
print(classification_report(y_test, y_pred))

['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
['malignant' 'benign']
dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename'])
.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from cen

/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


## 2-3 결론
Descision Tree : 0.91 Random Forest : 1.00 SVM :0.90 SGC : 0.89 Logistoc Regression : 0.95



# 총평
이번 노드에서는 머신러닝에 대한 기본개념을 배우는 단계가 아닐까 생각을 했었다. 
SVM, SGC, LR에 대해 들어는 봤어도 이를 잘 이해할 수 있을까 걱정이 앞섰다. 
하지만 최종 정리를 하며, Random Forest 성능을 통해 
이 모델의 성격이 앞으로 배우게 될 딥러닝의 전 단계로서 한번 짚고 넘어가는 단계였음을 알게 되었다.
여러 개의 의사결정 트리를 모아놓은 것으로 한 개의 Decision Tree의 단점을 극복하면서 
다른 모델의 성능보다 월등하게 높게 되는것이 결국 딥러닝의 가장 기본 개념이 아닐까 ..
노드의 내용 중 Random Forest를 집단지성이라고 표현을 했다. 
이번 aiffel교육도 집단지성을 통해 높은 성취효율을 추구하는것이 궁극적인 목표라 알고 있다.
본격적으로 과정이 궤도에 오르기전 이번 교육의 철학과 목표가 같음을 알 수 있었던 노드였다.